flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java

```diff
@@ -261,34 +261,69 @@ public void grantLeadership(UUID leaderSessionID) {
 
     @GuardedBy("lock")
     private void startJobMasterServiceProcessAsync(UUID leaderSessionId) {
-        sequentialOperation =
-                sequentialOperation.thenRun(
-                        () ->
-                                runIfValidLeader(
-                                        leaderSessionId,
-                                        ThrowingRunnable.unchecked(
-                                                () ->
-                                                        verifyJobSchedulingStatusAndCreateJobMasterServiceProcess(
-                                                                leaderSessionId)),
-                                        "verify job scheduling status and create JobMasterServiceProcess"));
-
+        boolean isValid;
+        synchronized (lock) {
+            isValid = isValidLeader(leaderSessionId);
+        }
+        if (isValid) {
+            sequentialOperation =
+                    sequentialOperation.thenCompose(
+                            unused ->
+                                    runIfValidLeader(
+                                                    leaderSessionId,
+                                                    () ->
+                                                            jobResultStore.hasJobResultEntryAsync(
+                                                                    getJobID()),
+                                                    "verify jbb result entry")
+                                            .handle(
+                                                    (hasJobResult, throwable) -> {
+                                                        if (hasJobResult == null) {
+                                                            return null;
+                                                        }
+                                                        if (hasJobResult) {
+                                                            runIfValidLeader(
```

Comment:
I guess we could remove code redundancy here by moving the `runIfValidLeader` out of the if clause. Alternatively, you could move the content of the two if/else blocks into their own private method with meaningful names (e.g. `handleJobAlreadyDoneIfValidLeader(leaderSessionID)` and `createNewJobMasterServiceProcessIfValidLeader(leaderSessionID)`) to improve this part of the code.

Reply:
I've added these two methods: `handleJobAlreadyDoneIfValidLeader(leaderSessionID)` and `createNewJobMasterServiceProcessIfValidLeader(leaderSessionID))`.